File: smalivm/src/main/java/org/cf/smalivm/ExceptionFactory.java
package org.cf.smalivm;

import org.cf.smalivm.opcode.Op;
import org.cf.util.ClassNameUtils;

import java.lang.reflect.Constructor;

public class ExceptionFactory {

    private final VirtualMachine vm;

    ExceptionFactory(VirtualMachine vm) {
        this.vm = vm;
    }

    public Throwable build(Op op, String className) {
        return build(op, className, null);
    }

    public Throwable build(Op op, Class<? extends Throwable> exceptionClass) {
        return build(op, exceptionClass, null);
    }

    public Throwable build(Op op, Class<? extends Throwable> exceptionClass, String message) {
        try {
            Constructor<? extends Throwable> ctor = exceptionClass.getDeclaredConstructor(String.class);
            ctor.setAccessible(true);
            return ctor.newInstance(message);
        } catch (Exception e) {
            e.printStackTrace();
        }

        return new Exception();
    }

    public Throwable build(Op op, String className, String message) {
        String binaryName = ClassNameUtils.internalToBinary(className);
        try {
            @SuppressWarnings("unchecked")
            Class<Throwable> exceptionClass = (Class<Throwable>) vm.getClassLoader().loadClass(binaryName);
            return build(op, exceptionClass, message);
        } catch (Exception e) {
            e.printStackTrace();
        }

        return new Exception();
    }
}

File: smalivm/src/main/java/org/cf/smalivm/ExceptionFactory.kt
import org.cf.smalivm.dex.SmaliClassLoader
import org.cf.smalivm.opcode.Op
import org.cf.util.ClassNameUtils

public class ExceptionFactory internal constructor(private val classLoader: SmaliClassLoader) {
    @JvmOverloads
    // TODO: op isn't needed? remove this? maybe this whole factory can get moved to node
    fun build(op: Op, exceptionClass: Class<out Throwable>, message: String? = null): Throwable {
        try {
            val ctor = exceptionClass.getDeclaredConstructor(String::class.java)
            ctor.isAccessible = true
            return ctor.newInstance(message)
        } catch (e: Exception) {
            e.printStackTrace()
        }
        return Exception()
    }

//    @JvmOverloads
//    fun build(op: Op, className: String, message: String? = null): Throwable {
//        val binaryName = ClassNameUtils.internalToBinary(className)
//        try {
//            val exceptionClass = classLoader.loadClass(binaryName) as Class<Throwable>
//            return build(op, exceptionClass, message)
//        } catch (e: Exception) {
//            e.printStackTrace()
//        }
//        return Exception()
//    }
}

